File: sdk/swift/Sources/HiTechCloudSDK/CephS3.swift
import Foundation

/// Ceph S3: S3 connection, credentials, usage, buckets, sub-users, key rotation
public class CephS3Resource {
    private let httpClient: HTTPClient
    public init(httpClient: HTTPClient) { self.httpClient = httpClient }

    /// GET /api/service/{id}/s3 - Get S3 connection info
    public func getConnectionInfo(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/s3", params: params)
    }

    /// GET /api/service/{id}/s3/credentials - Get S3 credentials with secret key
    public func getCredentials(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/s3/credentials", params: params)
    }

    /// GET /api/service/{id}/s3/usage - Get S3 usage
    public func getUsage(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/s3/usage", params: params)
    }

    /// GET /api/service/{id}/s3/metrics - Get S3 metrics
    public func getMetrics(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/s3/metrics", params: params)
    }

    /// GET /api/service/{id}/s3/buckets - List buckets
    public func listBuckets(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/s3/buckets", params: params)
    }

    /// POST /api/service/{id}/s3/buckets - Create bucket
    public func createBucket(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/s3/buckets", params: params)
    }

    /// DELETE /api/service/{id}/s3/buckets/{bucket} - Delete bucket
    public func deleteBucket(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.delete("/api/service/{id}/s3/buckets/{bucket}", params: params)
    }

    /// GET /api/service/{id}/s3/subusers - List sub-users
    public func listSubusers(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/s3/subusers", params: params)
    }

    /// POST /api/service/{id}/s3/subusers - Create sub-user
    public func createSubuser(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/s3/subusers", params: params)
    }

    /// DELETE /api/service/{id}/s3/subusers/{subuser} - Delete sub-user
    public func deleteSubuser(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.delete("/api/service/{id}/s3/subusers/{subuser}", params: params)
    }

    /// POST /api/service/{id}/s3/key - Rotate secret key
    public func rotateSecretKey(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/s3/key", params: params)
    }

}

File: sdk/kotlin/src/main/kotlin/vn/hitechcloud/sdk/CephS3.kt
package vn.hitechcloud.sdk

/** Ceph S3: S3 connection, credentials, usage, buckets, sub-users, key rotation */
class CephS3Resource(private val httpClient: HttpClient) {
    /** GET /api/service/{id}/s3 - Get S3 connection info */
    fun getConnectionInfo(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/s3", params)
    }

    /** GET /api/service/{id}/s3/credentials - Get S3 credentials with secret key */
    fun getCredentials(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/s3/credentials", params)
    }

    /** GET /api/service/{id}/s3/usage - Get S3 usage */
    fun getUsage(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/s3/usage", params)
    }

    /** GET /api/service/{id}/s3/metrics - Get S3 metrics */
    fun getMetrics(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/s3/metrics", params)
    }

    /** GET /api/service/{id}/s3/buckets - List buckets */
    fun listBuckets(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/s3/buckets", params)
    }

    /** POST /api/service/{id}/s3/buckets - Create bucket */
    fun createBucket(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/s3/buckets", params)
    }

    /** DELETE /api/service/{id}/s3/buckets/{bucket} - Delete bucket */
    fun deleteBucket(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.delete("/api/service/{id}/s3/buckets/{bucket}", params)
    }

    /** GET /api/service/{id}/s3/subusers - List sub-users */
    fun listSubusers(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/s3/subusers", params)
    }

    /** POST /api/service/{id}/s3/subusers - Create sub-user */
    fun createSubuser(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/s3/subusers", params)
    }

    /** DELETE /api/service/{id}/s3/subusers/{subuser} - Delete sub-user */
    fun deleteSubuser(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.delete("/api/service/{id}/s3/subusers/{subuser}", params)
    }

    /** POST /api/service/{id}/s3/key - Rotate secret key */
    fun rotateSecretKey(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/s3/key", params)
    }

}
